File: src/main/kotlin/com/vlnabatov/alabaster/extensions/annotation/XMLAnnotator.kt
package com.vlnabatov.alabaster.extensions.annotation

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.TEXT_ATTRIBUTES
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.BRACES
import com.intellij.psi.PsiElement
import com.intellij.psi.util.elementType
import com.intellij.psi.xml.XmlTokenType.*

class XMLAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        try {
            // delimiters
            if (element.elementType in setOf(
                    XML_EQ,
                    XML_ATTRIBUTE_VALUE_START_DELIMITER,
                    XML_ATTRIBUTE_VALUE_END_DELIMITER
                )
            ) {
                holder.newSilentAnnotation(TEXT_ATTRIBUTES).textAttributes(BRACES).create()
            }
        } catch (e: Exception) {
            /* Should not happen */
        }
    }
}

File: src/main/clojure/com/vlnabatov/alabaster/xml_annotator.clj
(ns com.vlnabatov.alabaster.xml-annotator
  (:require [com.vlnabatov.alabaster.util :refer [import-static]])
  (:import (com.intellij.lang.annotation AnnotationHolder)
           (com.intellij.lang.annotation HighlightSeverity)
           (com.intellij.openapi.editor DefaultLanguageHighlighterColors)
           (com.intellij.psi.xml XmlTokenType)
           (com.intellij.psi PsiElement))
  (:gen-class :implements [com.intellij.lang.annotation.Annotator]
              :name       com.vlnabatov.alabaster.extensions.annotation.XMLAnnotator
              :prefix     "-"))


(defn -annotate
  [_ ^PsiElement element ^AnnotationHolder holder]
  (when (#{XmlTokenType/XML_EQ XmlTokenType/XML_ATTRIBUTE_VALUE_START_DELIMITER
           XmlTokenType/XML_ATTRIBUTE_VALUE_END_DELIMITER}
         (.. element getNode getElementType))
    (do (-> holder
            (.newSilentAnnotation HighlightSeverity/INFORMATION)
            (.textAttributes DefaultLanguageHighlighterColors/NUMBER)
            (.create)))))
